drivers/ina3221/ina3221.c

```diff
@@ -0,0 +1,877 @@
+    if(flags)
+    {
+        status = _read_reg(dev, INA3221_REG_MASK_ENABLE, flags);
+        if(status < 0)
+        {
+            *flags = 0;
+            DEBUG("ina3221_read_shunt_sum_uV: Reading flags failed\n");
+        }
+        else
+        {
+            *flags &= INA3221_FLAGS_MASK;
+        }
+    }
+    return INA3221_OK;
+}
+
+int ina3221_read_shunt_uV(const ina3221_t* dev, ina3221_channel_t ch, int32_t* out_uV, uint16_t* flags)
+{
+    uint16_t reg_val;
+    int status = INA3221_OK;
+    int i, j;
+    for(i = 0, j = 0; i < INA3221_NUM_CH; i++)
+    {
+        if(ch & (1 << i))
+        {
+            status = _read_reg(dev, _chi[i].chi_reg_shunt, &reg_val);
+            if(status < 0)
+            {
+                break;
+            }
+            out_uV[j++] = reg_val_to_shunt_voltage_uV(reg_val);
+        }
+    }
+    if(j && flags)
+    {
+        status = _read_reg(dev, INA3221_REG_MASK_ENABLE, flags);
+        if(status < 0)
+        {
+            *flags = 0;
+            DEBUG("ina3221_read_shunt_uV: Reading flags failed\n");
+        }
+        else
+        {
+            *flags &= INA3221_FLAGS_MASK;
+        }
+    }
+    return j ? j : status;
+}
+
+int ina3221_read_bus_mV(const ina3221_t* dev, ina3221_channel_t ch, int16_t* out_mV, uint16_t* flags)
+{
+    uint16_t reg_val;
+    int status = INA3221_OK;
+    int i, j = 0;
+    for(i = 0; i < INA3221_NUM_CH; i++)
+    {
+        if(ch & (1 << i))
+        {
+            status = _read_reg(dev, _chi[i].chi_reg_bus, &reg_val);
+            if(status < 0)
+            {
+                break;
+            }
+            out_mV[j++] = reg_val_to_bus_voltage_mV(reg_val);
+        }
+    }
+    if(j && flags)
+    {
+        status = _read_reg(dev, INA3221_REG_MASK_ENABLE, flags);
+        if(status < 0)
+        {
+            *flags = 0;
+            DEBUG("ina3221_read_bus_mV: Reading flags failed\n");
+        }
+        else
+        {
+            *flags &= INA3221_FLAGS_MASK;
+        }
+    }
+    return j ? j : status;
+}
+
+int ina3221_calculate_current_uA(const ina3221_t* dev, ina3221_channel_t ch, int32_t* in_uV, int32_t* out_uA)
+{
+    int i, j = 0;
+    for(i = 0; i < INA3221_NUM_CH; i++)
+    {
+        if(ch & (1 << i))
+        {
+            out_uA[j++] = in_uV[i] * 1000 / dev->params.rshunt_mOhm[i];
+        }
+    }
+    return j;
+}
+
+int ina3221_calculate_power_uW(int16_t* in_mV, int32_t* in_uA, uint8_t num, int32_t* out_uW)
+{
+    if(num > INA3221_NUM_CH)
+    {
+        return -ERANGE;
+    }
+    int i;
+    for(i = 0; i < num; i++)
+    {
+        /* max 26V bus voltage */
+        /* (2^31)-1 resolution; 2.147483647 Watt in Nanowatt resolutiona */
+        /* 2.147483647 / 26000 = 82595.525 */
+        if(in_uA[i] < (82596 - 500))
+        {
+            out_uW[i] = (in_uA[i] * in_mV[i] + 500) / 1000;
+        }
+        else
+        {
+            out_uW[i] = (in_uA[i] + 500) / 1000 * in_mV[i];
+        }
+    }
+    return i;
+}
```

Comment:
The last line should be terminated by exactly one newline. Here the last byte of the file is `}`. Most editors can be configured to add the terminating new line when saving, and most even do so by default.